test/charms/test_secrets/src/charm.py

```
@@ -80,6 +82,53 @@ def add_with_meta(self, event: ops.ActionEvent):
         }
         event.set_results(cast('dict[str, Any]', result))
 
+    def set_secret_flow(self, event: ops.ActionEvent):
+        result: Result = {}
+        contentiter = iter(contents())
+        labeliter = iter(labels())
+        descriptioniter = iter(descriptions())
+        expireiter = iter(expires())
+        rotateiter = iter(rotates())
+
+        try:
+            secretid = event.params.get('secretid')
+            secretlabel = event.params.get('secretlabel')
+
+            for field in event.params['flow'].split(','):
+                assert secretid or secretlabel
+                assert not (secretid and secretlabel)
+                secret = self.model.get_secret(
+                    **({'id': secretid} if secretid else {}),
+                    **({'label': secretlabel} if secretlabel else {}),
+                )
+
+                match field:
+                    case 'content':
+                        secret.set_content(next(contentiter))
+                    case 'label':
+                        new_label = next(labeliter)
+                        secret.set_info(label=new_label)
+                        if secretlabel:
+                            # So that we can find the secret again
+                            secretlabel = new_label
+                    case 'description':
+                        secret.set_info(description=next(descriptioniter))
+                    case 'expire':
+                        secret.set_info(expire=next(expireiter))
+                    case 'rotate':
+                        secret.set_info(rotate=next(rotateiter))
+                    case _:
+                        raise ValueError(f'Unsupported {field=}')
+
+            if not secretid:
+                secretid = self.model.get_secret(label=secretlabel).get_info().id
+            result['secretid'] = secretid
+            result['after'] = self._snapshot(secretid)
+        except Exception as e:
+            result['exception'] = str(e)
```

How about using [event.fail](https://documentation.ubuntu.com/ops/latest/reference/ops/#ops.ActionEvent.fail) here, since the tests first `assert` that we don't have an exception?

---

Do we need the try/except at all? It seems simpler to let it fail, which for Scenario would bubble up the exception wrapped in UncaughtCharmError and the test would (correctly) fail, and for Jubilant the charm would go to error state (which I assume means again the test correctly fails).

+1 on event.fail if we *expect* a failure, but I don't think we should catch unexpected ones.